crates/bws/src/main.rs

```diff
@@ -404,6 +409,102 @@ async fn process_commands() -> Result<()> {
             }
         }
 
+        Commands::Run {
+            command,
+            shell,
+            no_inherit_env,
+            project_id,
+        } => {
+            let shell = match std::env::consts::OS {
+                "windows" => shell.unwrap_or_else(|| "powershell".to_string()),
+                _ => shell.unwrap_or_else(|| "sh".to_string()),
+            };
+
+            if which(&shell).is_err() {
+                eprintln!("Error: shell '{}' not found", shell);
+                std::process::exit(1);
+            }
+
+            let user_command = if command.is_empty() {
+                if atty::is(Stream::Stdin) {
+                    eprintln!("{}", Cli::command().render_help().ansi());
+                    std::process::exit(1);
+                }
+
+                let mut buffer = String::new();
+                std::io::stdin().read_to_string(&mut buffer)?;
+                buffer
+            } else {
+                command.join(" ")
+            };
+
+            if user_command.is_empty() {
+                let mut cmd = Cli::command();
+                eprintln!("{}", cmd.render_help().ansi());
+                std::process::exit(1);
+            }
+
+            let res = if let Some(project_id) = project_id {
+                client
+                    .secrets()
+                    .list_by_project(&SecretIdentifiersByProjectRequest { project_id })
+                    .await?
+            } else {
+                client
+                    .secrets()
+                    .list(&SecretIdentifiersRequest { organization_id })
+                    .await?
+            };
+
+            let secret_ids = res.data.into_iter().map(|e| e.id).collect();
+            let secrets = client
+                .secrets()
+                .get_by_ids(SecretsGetRequest { ids: secret_ids })
+                .await?
+                .data;
+
+            let environment = secrets
+                .into_iter()
+                .map(|s| (s.key, s.value))
+                .collect::<std::collections::HashMap<String, String>>();
+
+            for key in environment.keys() {
+                if !is_valid_posix_name(key) {
```

Comment:
will shell start with invalid environment variables ? don't we need to filter those out ?

Reply:
> 1. Will this work with quoted args and multiline ?
> ```
> bws run docker run --platform "linux/amd64" --rm --name "ubuntu-1" ubuntu echo 'hello'
> 
> bws run if [ "$ENV1" = "1234" ]; then
> echo "hello";
> fi
> ```

Great questions!

In the first case, I'd use an "end of arguments" (`--`) operator to construct a command like that to ensure that the shell is passing all of the arguments to our command: `bws run -- docker run --platform "linux/amd64" --rm --name "ubuntu-1" ubuntu echo 'hello'`.

In the case of the multi-line `if` statement, I'm not sure if there's a way for `bws` to handle that, as the first new-line passed to the shell would immediately attempt to execute `bws run if [ "$ENV1" = "1234" ]; then` on its own. I'm unsure if anything other than a shell builtin could trigger the interactive multiline input.

I wouldn't generally use the `run` command to execute ad-hoc shell commands like that, but if you wanted to, you could wrap it in single-quotes to avoid having the shell executing or interpolating anything before it's passed to our command. Something like the following would work:
```bash
bws run -- 'if [ "$ENV1" = "1234" ]; then
echo "hello";
fi'
```

> 2. Is creating the environment variables secrets with `key` good idea, since it is not guaranteed to be unique within one and across multiple projects and does not follow posix env format ?

I would prefer keeping the default behavior to use the key names, as it greatly improves usability and is in-line with CLIs that have similar functionality. Making UUIDs the default would require more effort for users to use the `run` command to replace things like local env files, where the environment variables wouldn't have a Bitwarden-generated secret UUID to reference.

>  Maybe it should be uuid by default (with replaced - characters to underscores _) with cli option like --use-secret-key to additionally expand to env variable by key (to be more user friendly, maybe even prefixed by project name or id ?)

What would you think of a `BWS_USE_UNIQUE_NAMES=1` environment variable that could be set that would ensure that all secrets that are set get the secret UUID (in snake_case) appended to them? So, if you're running a script where you have secrets with identical names, you can ensure uniqueness. We could possibly add an equivalent argument (`--use-unique-names`?), but I feel like you would be much more likely to need that functionality on a per-script/session basis, rather than as a one-off. Either way, I think having something like that would help people that can't guarantee the uniqueness of their secret names.

Reply:
> will shell start with invalid environment variables ? don't we need to filter those out ?

It will set the environment variables, even if they're not POSIX-compliant, yes. You'll get warnings printed to stderr stating that you have some non-POSIX-compliant key names, but they are technically set correctly by the `run` command. I figured that if someone were to executing a binary/script that had less restrictive naming conventions for environment variables, we could leave them set. For example:

```bash
bws run -- 'python -c "import os; print(os.getenv("1 non-compliant keyname"))"'
```

would work, since `getenv` doesn't seem to require POSIX compliance. I imagine the same is possible for equivalent libs in other languages.

I'm open to un-setting them, but I'm not sure how people might feel about that if they're used to not being limited by POSIX-compliance.

Reply:
> --use-unique-name

I like this, being explicit in the cli options is better. 
And like you mentioned, makes the solution more future proof, when users wants to do things by Id.

Reply:
> I'm open to un-setting them, but I'm not sure how people might feel about that if they're used to not being limited by POSIX-compliance.

I am fine either way, just want to avoid `bws run` not working because of some special characters in the secret key.